buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java

```diff
@@ -363,11 +363,18 @@ static void setBytes(AbstractByteBuf buf, long addr, int index, ByteBuffer src)
             // Copy from direct memory
             long srcAddress = PlatformDependent.directBufferAddress(src);
             PlatformDependent.copyMemory(srcAddress + src.position(), addr, src.remaining());
-        } else {
+        } else if (src.hasArray()) {
             // Copy from array
             PlatformDependent.copyMemory(src.array(), src.arrayOffset() + src.position(), addr, length);
+        } else {
+            byte[] buffer = new byte[length];
```

Comment:
I think best would be too allocate a heap buffer from a ByteBufAllocator and use this to do the copy to allow to reduce garbage if a PooledByteBufAllocator is used. Like here: https://github.com/jakubjanecek/netty/blob/fix/SetBytesOnReadOnlyByteBuffer/buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java#L269